File: lab3/src/geometry/GeometryMath.java
package geometry;

import shapes.Shape;

public class GeometryMath {

    public static final double PI = Math.PI;
    
    /**
<<<<<<< HEAD
     * Перевод величины угла
     * @param degrees - угол в градусах
=======
     * Перевод градусов в радианы
     * @param degrees угол в градусах
>>>>>>> b67dc2cd44ad23232b79af10259c81a92da0a34c
     * @return угол в радианах
     */
    public static double degreesToRadian(int degrees) {
        return (degrees % 360) * PI / 180;
    }

    /**
     * Сравнение фигур по площади
     */
    public static int compareByArea(Shape lhs, Shape rhs) {
        return Double.compare(lhs.getArea(), rhs.getArea());
    }

<<<<<<< HEAD

    /**
     * Определение пересечения фигур
=======
    /**
     * Проверка пересечения Фигур
>>>>>>> b67dc2cd44ad23232b79af10259c81a92da0a34c
     */
    public static boolean isIntersect(Shape lhs, Shape rhs) {
        var lhsEdges = lhs.getEdgeLines();
        var rhsEdges = rhs.getEdgeLines();

        for (var i = 0; i < lhsEdges.length; i++) {
            var l1 = lhsEdges[i];

            for (var j = 0; j < rhsEdges.length; j++) {
                var l2 = rhsEdges[j];

                if (l1.isIntersecting(l2)) {
                    return true;
                }
            }
        }

        return false;
    }
    
    /**
<<<<<<< HEAD
     * Определение включения точки в фигуру
     * Алгоритм строит линии с началом в точке P, и концами на точках (вершинах) Shape
     * Если угол между всеми этими линиями равен 360 -> точка включается в фигуру
=======
     * Проверка включения ТОЧКИ в фигуру
>>>>>>> b67dc2cd44ad23232b79af10259c81a92da0a34c
     */
    public static boolean isIncludes(Point p, Shape shape) {
        var points = shape.getPoints();

        for (var i = 0; i < points.length; i++) {
            if (p.equals(points[i])) {
                return true;
            }
        }

        var checkLinesCount = points.length + 1;
        var checkLines = new Line[checkLinesCount];

        for (var i = 0; i < checkLinesCount - 1; i++) {
            checkLines[i] = new Line(p, points[i]);
        }
        checkLines[checkLinesCount - 1] = checkLines[0];

        double angle = 0;
        for (var i = 0; i < checkLinesCount - 1; i++) {
            angle += Line.angle(checkLines[i], checkLines[i + 1]);
        }

        return angle == 2 * PI;
    }

    /**
<<<<<<< HEAD
     * Определение включения фигуры в другой
=======
     * Проверка включения одной фигуры в другой
>>>>>>> b67dc2cd44ad23232b79af10259c81a92da0a34c
     */
    public static boolean isIncludes(Shape lhs, Shape rhs) {
        var points = rhs.getPoints();

        for (var i = 0; i < points.length; i++) {
            if (isIncludes(points[i], lhs) == false) {
                return false;
            }
        }

        return true;
    }
}

File: lab5/src/geometry/GeometryMath.kt
package geometry

import shapes.Shape

object GeometryMath {

    const val PI = Math.PI

    /**
     * Перевод величины угла
     * @param degrees - угол в градусах
     * @return угол в радианах
     */
    fun degreesToRadian(degrees: Int): Double {
        return (degrees % 360) * PI / 180
    }

    /**
     * Сравнение фигур по площади
     */
    fun compareByArea(lhs: Shape, rhs: Shape): Int {
        return lhs.getArea().compareTo(rhs.getArea())
    }

    /**
     * Определение пересечения фигур
     */
    fun isIntersect(lhs: Shape, rhs: Shape): Boolean {
        val lhsEdges = lhs.getEdgeLines()
        val rhsEdges = rhs.getEdgeLines()

        for (l1 in lhsEdges) {
            for (l2 in rhsEdges) {
                if (l1.isIntersecting(l2)) {
                    return true
                }
            }
        }

        return false
    }

    /**
     * Определение включения точки в фигуру
     * Алгоритм строит линии с началом в точке P, и концами на точках (вершинах) Shape
     * Если угол между всеми этими линиями равен 360 -> точка включается в фигуру
     */
    fun isIncludes(p: Point, shape: Shape): Boolean {
        val points = shape.getPoints()

        for (point in points) {
            if (p == point) {
                return true
            }
        }

        val checkLinesCount = points.size + 1
        val checkLines = arrayOfNulls<Line>(checkLinesCount)

        for (i in 0..< checkLinesCount - 1) {
            checkLines[i] = Line(p, points[i])
        }
        checkLines[checkLinesCount - 1] = checkLines[0]

        var angle = 0.0
        for (i in 0..< checkLinesCount - 1) {
            angle += Line.angle(checkLines[i]!!, checkLines[i + 1]!!)
        }

        return angle == 2 * PI
    }

    /**
     * Определение включения фигуры в другую
     */
    fun isIncludes(lhs: Shape, rhs: Shape): Boolean {
        val points = rhs.getPoints()

        for (point in points) {
            if (!isIncludes(point, lhs)) {
                return false
            }
        }

        return true
    }
}
